File: src/shady/generator_visit.c
#include "generator.h"

void generate(Growy* g, json_object* src) {
    generate_header(g, src);

    json_object* nodes = json_object_object_get(src, "nodes");
    growy_append_formatted(g, "void visit_node_operands(Visitor* visitor, NodeClass exclude, const Node* node) {\n");
    growy_append_formatted(g, "\tswitch (node->tag) { \n");
    assert(json_object_get_type(nodes) == json_type_array);
    for (size_t i = 0; i < json_object_array_length(nodes); i++) {
        json_object* node = json_object_array_get_idx(nodes, i);
        String name = json_object_get_string(json_object_object_get(node, "name"));
        String snake_name = json_object_get_string(json_object_object_get(node, "snake_name"));
        void* alloc = NULL;
        if (!snake_name) {
            alloc = snake_name = to_snake_case(name);
        }
        growy_append_formatted(g, "\tcase %s_TAG: {\n", name);
        json_object* ops = json_object_object_get(node, "ops");
        if (ops) {
            assert(json_object_get_type(ops) == json_type_array);
            growy_append_formatted(g, "\t\t%s payload = node->payload.%s;\n", name, snake_name);
            for (size_t j = 0; j < json_object_array_length(ops); j++) {
                json_object* op = json_object_array_get_idx(ops, j);
                String op_name = json_object_get_string(json_object_object_get(op, "name"));
                String class = json_object_get_string(json_object_object_get(op, "class"));
                if (!class || strcmp(class, "string") == 0)
                    continue; // skip 'string' class and POD operands
                String class_cap = capitalize(class);
                bool list = json_object_get_boolean(json_object_object_get(op, "list"));
                bool ignore = json_object_get_boolean(json_object_object_get(op, "ignore"));
                if (!ignore) {
                    growy_append_formatted(g, "\t\tif ((exclude & Nc%s) == 0)\n", class_cap);
                    if (list)
                        growy_append_formatted(g, "\t\t\tvisit_ops(visitor, Nc%s, \"%s\", payload.%s);\n", class_cap, op_name, op_name);
                    else
                        growy_append_formatted(g, "\t\t\tvisit_op(visitor, Nc%s, \"%s\", payload.%s);\n", class_cap, op_name, op_name);
                }
                free(class_cap);
            }
        }
        growy_append_formatted(g, "\t\tbreak;\n");
        growy_append_formatted(g, "\t}\n", name);
        if (alloc)
            free(alloc);
    }
    growy_append_formatted(g, "\t\tdefault: assert(false);\n");
    growy_append_formatted(g, "\t}\n");
    growy_append_formatted(g, "}\n\n");
}

File: src/shady/generator_visit.py
from src.generator.generator import *
from src.generator.generator_common import *
from src.generator.generator_main import main

def generate(g, src):
    generate_header(g, src)

    nodes = src["nodes"]
    g.g += "void shd_visit_node_operands_generated(Visitor* visitor, NodeClass exclude, const Node* node) {\n"
    g.g += "\tswitch (node->tag) { \n"
    assert type(nodes) is list
    for node in nodes:
        name = node["name"]
        snake_name = node["snake_name"]

        g.g += f"\tcase {name}_TAG: {{\n"
        ops = node.get("ops")
        if ops is not None:
            assert type(ops) is list
            g.g += f"\t\tSHADY_UNUSED {name} payload = node->payload.{snake_name};\n"
            for op in ops:
                op_name = op["name"]
                is_list = op.get("list")
                ignore = op.get("ignore")
                if (ignore):
                    continue
                clazz = op.get("class")
                if clazz is None:
                    continue
                elif clazz == "string":
                    continue
                else:
                    class_cap = capitalize(clazz)
                    g.g += f"\t\tif ((exclude & Nc{class_cap}) == 0)\n"
                    if is_list:
                        g.g += f"\t\t\tshd_visit_ops(visitor, Nc{class_cap}, \"{op_name}\", payload.{op_name});\n"
                    else:
                        g.g += f"\t\t\tshd_visit_op(visitor, Nc{class_cap}, \"{op_name}\", payload.{op_name}, 0);\n"
        g.g += f"\t\tbreak;\n"
        g.g += f"\t}}\n"
    g.g += "\t\tdefault: assert(false);\n"
    g.g += "\t}\n"
    g.g += "}\n\n"

main(generate)